### PR TITLE
perf: Improve arrow head sizing

### PR DESCRIPTION
**Pull request by lily-ge**
making it proportional to the length of the last segment of the arrow

### CODE REVIEW

**Comment by lily-ge** [comment]
This change improves the arrow head sizing as described in this issue https://github.com/excalidraw/excalidraw/issues/3263

**Review by dwelle — APPROVED** [review]
:+1: for me.

PR:

image

production: 

image

**Comment by lily-ge** [comment]
Thank you!

**Comment by wietsevenema** [comment]
This is awesome, thank you! 🥰

File: src/element/bounds.ts [hunk 1]
```diff
@@ -220,12 +220,11 @@ export const getArrowheadPoints = (
 
   const length = element.points.reduce((total, [cx, cy], idx, points) => {
     const [px, py] = idx > 0 ? points[idx - 1] : [0, 0];
-    return total + Math.hypot(cx - px, cy - py);
+    return Math.hypot(cx - px, cy - py);
   }, 0);
```
Comment (dwelle) [inline src/element/bounds.ts:224]: can we remove the reduce and just use the last two points for better perf?

**Review by dwelle — CHANGES_REQUESTED** [review]

File: src/element/bounds.ts (hunk 1 above)
Comment (lily-ge) [reply src/element/bounds.ts]: Yes! The changes have been pushed

**Review by ad1992 — APPROVED** [review]

**Comment by dwelle** [comment]
I came across to cases which we may wanna figure out first. /cc @wietsevenema 

1. Does it make sense to adaptively change size of non-arrow endpoints, too?
	
	production:

	image

	PR:

	image

	I think I prefer the production behavior.

2. The new algorithm incorrectly doesn't account for start arrowhead and computes the size from the last segment as well. Results in this:
	
	image

	It could be considered a feature (both ends are the same size), or a bug. Thougts?

**Comment by wietsevenema** [comment]
Here's one example where adaptive sizing does not make sense for the bar arrow head:
image

I've never used the dotted arrow head, but I think it would make sense to scale it (I don't have an example)

I think scaling both arrow heads is the least surprising behavior. It's a feature.

**Comment by wietsevenema** [comment]
Hi, I wanted to drop another real-world example in to this thread to keep the discussion going :)
image

**Comment by dwelle** [comment]
Forgot to get back to this. So I propose we autosize only arrow arrowheads for now, how does that sound?

@wietsevenema as for your last example, can you copy/paste it into this PR's preview and show how it looks when the arrow is resized? Won't it make the diagram look a bit weird, compared to the 2nd arrow?

**Comment by steveruizok** [comment]
image

Short arrows that are mostly in the same direction can produce small arrowheads even though the line maybe shouldn't? On one hand this could be a sneaky way to control the size of the arrowhead, but if we wanted to fix it we could throw some dot products in:

Kapture 2021-05-11 at 12 11 52

```ts
let length = 0
let pu = [0, 0];
for (let i = 0; i < element.points.length; i++) {
  const [px, py] = element.points[i - 1] || [0, 0];
  const [cx, cy] = element.points[i];
  
  const len = Math.hypot(cx - px, cy - py);
  const cu = [(cx - px) / len, (cy - py) / len];
  
  if (i > 1) {
    const dpr = pu[0] * cu[0] + pu[1] * cu[1];
    length = length * Math.max(0, dpr * dpr * dpr);
  }
  
  length += len;
  pu = cu;
}
```

**Review by dwelle — APPROVED** [review]

**Comment by wietsevenema** [comment]
Example with the diagram I shared earlier. This is great. And I really like that trick of using a small segment at the end to control arrow head size :)

image

**Comment by steveruizok** [comment]
Merge?

**Comment by dwelle** [comment]
Oops. This got forgot! Thanks @lily-ge & everyone!

**Pull request merged**